src/acl/FilledChecklist.h

```diff
@@ -32,13 +32,27 @@ class ConnStateData;
  */
 class ACLFilledChecklist: public ACLChecklist
 {
-    CBDATA_CLASS(ACLFilledChecklist);
+    CBDATA_CLASS_WITH_MAKE(ACLFilledChecklist);
 
 public:
+    /// Unlike regular Foo::Pointer types, this smart pointer is meant for use
+    /// during checklist configuration only, when it provides exception safety.
+    /// Any other/long-term checklist storage requires CbcPointer or equivalent.
+    using MakingPointer = std::unique_ptr<ACLFilledChecklist>;
+
     ACLFilledChecklist();
     ACLFilledChecklist(const acl_access *, HttpRequest *);
     ~ACLFilledChecklist() override;
 
+    /// Creates an ACLFilledChecklist object with given constructor arguments.
+    /// Callers are expected to eventually proceed with NonBlockingCheck().
+    static MakingPointer Make(const acl_access *a, HttpRequest *r) { return MakingPointer(new ACLFilledChecklist(a, r)); }
+
+    /// \copydoc ACLChecklist::nonBlockingCheck()
+    /// This public nonBlockingCheck() wrapper should be paired with Make(). The
+    /// pair prevents exception-caused Checklist memory leaks in caller code.
+    static void NonBlockingCheck(MakingPointer &&p, ACLCB *cb, void *data) { p->nonBlockingCheck(cb, data); (void)p.release(); }
```

Comment:
Earlier code simply said `p->release()->nonBlockingCheck()`, but that sequence would leak the checklist object if nonBlockingCheck() throws. Today, nonBlockingCheck() may call the callback directly, so the exception may be thrown not just by Checklist code, but by the callback code as well.

I hope the current version is safer, even though `p` becomes dangling between nonBlockingCheck() and release() calls. Please double check.

Reply:
Imagine a scenario when nonBlockingChecks() successfully starts off some non-blocking checks (passing the list to them) and then throws, allowing unique_ptr to delete the list. When one of those previously started ACLs completes, it calls resumeNonBlockingCheck() with the already deleted list. Is it possible? Or throwing in nonBlockingCheck() would abort those ACLs as well?

Reply:
> When one of those previously started ACLs completes, it calls resumeNonBlockingCheck() with the already deleted list. Is it possible?

I hope the correct answer is "no, it is not possible" because deleted checklist will not pass cbdata checks that are performed for asychronous callbacks that have a checklist object as a parameter. For example, this callback will not be called:

```C++
/// Called when an async lookup returns
void
ACLExternal::LookupDone(void *data, const ExternalACLEntryPointer &result)
{
    ACLFilledChecklist *checklist = Filled(static_cast<ACLChecklist*>(data));
    checklist->extacl_entry = result;
    checklist->resumeNonBlockingCheck();
}
```

... because of this cbdata check in externalAclHandleReply:

```C++
        if (state->callback && cbdataReferenceValidDone(state->callback_data, &cbdata))
            state->callback(cbdata, entry);
```

Reply:
Ok, I looked through other callers - all of them call cbdataReferenceValidDone().